src/Cli/dotnet/Commands/Run/VirtualProjectBuildingCommand.cs

```diff
@@ -847,7 +847,28 @@ public static void RemoveDirectivesFromFile(ImmutableArray<CSharpDirective> dire
 
     public static bool IsValidEntryPointPath(string entryPointFilePath)
     {
-        return entryPointFilePath.EndsWith(".cs", StringComparison.OrdinalIgnoreCase) && File.Exists(entryPointFilePath);
+        if (!File.Exists(entryPointFilePath))
+        {
+            return false;
+        }
+
+        if (entryPointFilePath.EndsWith(".cs", StringComparison.OrdinalIgnoreCase))
+        {
+            return true;
+        }
+
+        // Check if the first two characters are #!
+        try
+        {
+            using var stream = File.OpenRead(entryPointFilePath);
```

Comment:
In the case there is a BOM this is going to fail right? Think we need to use a reader API that is BOM aware and grab the first two `char` from it to compare to the prefix.

Reply:
Ugh 💣

Reply:
Interesting. Would not have thought that. Let's match that behavior then: aka keep as is.